### ISSUE TITLE
cmake is overriding distro hardening build flags

### ISSUE THREAD

**Issue by space88man** [kamailio/kamailio#4082]
### Description
ping @xkaraman 

cmake is  overriding distro build flags

Here is typical CFLAGS for EL distros

```
export CFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'

## try to build
cd build/
make
[  1%] Building C object src/CMakeFiles/kamailio.dir/lib/srdb1/db_row.c.o
In file included from /usr/include/bits/libc-header-start.h:33,
                 from /usr/include/string.h:26,
                 from /home/centos/src/kamailio-5.8/src/lib/srdb1/../../core/str.h:26,
                 from /home/centos/src/kamailio-5.8/src/lib/srdb1/db_con.h:33,
                 from /home/centos/src/kamailio-5.8/src/lib/srdb1/db_val.h:40,
                 from /home/centos/src/kamailio-5.8/src/lib/srdb1/db_row.h:37,
                 from /home/centos/src/kamailio-5.8/src/lib/srdb1/db_row.c:33:
/usr/include/features.h:412:4: warning: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Wcpp]
  412 | #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
```

It is probably due to `cmake/compiler-specific.cmake`:
```
      target_compile_options(
        common INTERFACE -O0
                         # <$<$<BOOL:${PROFILE}>:-pg>
      )

```

Packaging for ELx recommends using the default flags as far as possible - for EL9 included here for reference
```
CFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
CXXFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
FCFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
FFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
LDFLAGS='-Wl,-z,relro -Wl,--as-needed -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 '
```

ELx packaging typically involves this dance:

```
# in practice RPM build tools have macros for all the following steps
# use standard distro flags
export CFLAGS=.... # well-known distro CFLAGS
export CXXFLAGS=... # well-known distro CXXFLAGS
export LDFLAGS=.... # well-known distro CXXFLAGS

./configure 
```
.deb packaging also has a similar ritual like
```
      # do this first before running configure/cmake
      # support http://wiki.debian.org/Hardening for >=wheezy
      CFLAGS+=	$(shell dpkg-buildflags --get CFLAGS)
      CPPFLAGS+=	$(shell dpkg-buildflags --get CPPFLAGS)
      LDFLAGS+=	$(shell dpkg-buildflags --get LDFLAGS)
      LDLIBS+=	$(shell dpkg-buildflags --get LDLIBS)
```
Is there a way we can respect distro build flags with cmake?

Can we detect CFLAGS/LDFLAGS in the environment and use that instead?

**Comment by space88man** [comment]
Suggestion for `cmake/compiler-specific.cmake` - allow user to override CFLAGS/LDFLAGS:

```
if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")

  if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
    target_compile_definitions(common INTERFACE CC_GCC_LIKE_ASM)

    if(NOT DEFINED ENV{CFLAGS})
      target_compile_options(
        common INTERFACE -O0
                         # <$<$<BOOL:${PROFILE}>:-pg>
      )

      target_compile_options(
        common
        INTERFACE -Wall -funroll-loops -Wcast-align
                  -Werror=implicit-function-declaration -Werror=implicit-int
      )

      # If GCC version is greater than 4.2.0, enable the following flags
      if(CMAKE_C_COMPILER_VERSION VERSION_GREATER 4.2.0)
        target_compile_options(
          common INTERFACE -m64 -minline-all-stringops -falign-loops
                           -ftree-vectorize -fno-strict-overflow -mtune=generic
        )
      endif()
    endif()
  elseif(CMAKE_C_COMPILER_ID STREQUAL "Clang")
    target_compile_definitions(common INTERFACE CC_GCC_LIKE_ASM)
    if(NOT DEFINED ENV{CFLAGS})
      target_compile_options(common INTERFACE -m64)
    endif()
    if(NOT DEFINED ENV{LDFLAGS})
      target_link_options(common INTERFACE -m64)
    endif()
  endif()

# etc etc

```

**Comment by xkaraman** [comment]
Hey @space88man,

We had this discussion with @linuxmaniac as well. In https://github.com/kamailio/kamailio/issues/4053#issuecomment-2545146514, you can find a way to provide the `CFLAGS` and the `LDFLAGS`. In [Environment Variables for Languages](https://cmake.org/cmake/help/latest/manual/cmake-env-variables.7.html#id4|), you can find all the `ENV` flags that you can pass to `cmake`. These `FLAGS` are respected when called on the **first time** of `cmake` configuration. We can also define/modify [CMAKE_[LANG]_FLAGS](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS.html#variable:CMAKE_%3CLANG%3E_FLAGS) after the first time config phase.

Note though, this will initialize the `FLAGS` with your provided ones and append to them the project's one found in `target_compile_options`. Is this ok, or do you need to completely override them and only use whatever you provide?

**Comment by space88man** [comment]
@xkaraman, I definitely want to keep all the `-DHAVE_XXX` flags deduced by the build system; however the compiler options like `-O0 -mtune=generic` need to be completely overridden.

I tried just defining `CFLAGS="-O2 -g …"` but that got appended/prepended (?) to the CFLAGS  so RHELs `FORTIFY_SOURCE` preference didn't work - it looks like `-O0` (from current cmake) "won".

Summary: need to keep preprocessor defines, but override compiler options.

**Comment by xkaraman** [comment]
@space88man Hmm, i just read through https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html. Since, the one you define is passed first (prepended) followed by the project options (each `target_compile_options` appends to the last ones), the last `-O` wins the race. 

I think since the default option of GCC is `-O0` if none is passed, I think it's safe to remove it from the `compiler-specific.cmake`, so if one want to alter the optimization level, it's safe to do it via the above `CFLAGS="-O2"` method.

**Comment by space88man** [comment]
Closed with 5d4002415f.
